### PR TITLE
fix(crd): perform pre-inflight checks for m-apiserver

### PR DESCRIPTION
**Pull request by vishnuitta**
Signed-off-by: Vitta <vitta@mayadata.io>

**What this PR does / why we need it**:
This PR adds pre-inflight checks for maya-apiserver. Currently, NDM related CRDs checks are made.
Also, for pre-upgrade failures, this PR does fatal.

manual testing done.
TODO: BDDs are pending

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:

**Checklist:**
- [ ] Fixes #<issue number>
- [ ] Labelled this PR & related issue with `documentation` tag
- [ ] PR messages has document related information
- [ ] Labelled this PR & related issue with `breaking-changes` tag
- [ ] PR messages has breaking changes related information
- [ ] Labelled this PR & related issue with `requires-upgrade` tag
- [ ] PR messages has upgrade related information
- [ ] Commit has unit tests
- [ ] Commit has integration tests

### CODE REVIEW

File: cmd/maya-apiserver/app/command/start.go [hunk 1]
```diff
@@ -126,6 +129,22 @@ func NewCmdStart() *cobra.Command {
...
 }
 
+func performPreInflightChecks() error {
+	return checkForNDMrelatedCRDs()
+}
+
+func checkForNDMrelatedCRDs() error {
+	_, err := bdc.NewKubeClient().List(metav1.ListOptions{})
+	if err != nil {
+		return fmt.Errorf("precheck for bdc failed: %v", err)
```
Comment (mynktl) [inline cmd/maya-apiserver/app/command/start.go]: Let's use `error.Errorf` here. We are using `fmt` for string related operation only.

File: cmd/maya-apiserver/app/command/start.go [hunk 2]
```diff
@@ -126,6 +129,22 @@ func NewCmdStart() *cobra.Command {
 	return cmd
 }
 
+func performPreInflightChecks() error {
+	return checkForNDMrelatedCRDs()
+}
+
+func checkForNDMrelatedCRDs() error {
+	_, err := bdc.NewKubeClient().List(metav1.ListOptions{})
```
Comment (mittachaitu) [inline cmd/maya-apiserver/app/command/start.go]: Instead of `CR` we need to check for `CRD`

File: cmd/maya-apiserver/cstor-operator/spc/runner.go [hunk 3]
```diff
@@ -31,7 +31,12 @@ import (
 // is closed, at which point it will shutdown the workqueue and wait for
 // workers to finish processing their current work items.
 func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
-	defer runtime.HandleCrash()
+	var err error
+	defer func() {
```
Comment (mittachaitu) [inline cmd/maya-apiserver/cstor-operator/spc/runner.go]: Why we are not doing fatal in caller of Run

File: cmd/maya-apiserver/app/command/start.go (hunk 2 above)
Comment (mittachaitu) [reply cmd/maya-apiserver/app/command/start.go]: Ignore above comment

Comment (vishnuitta) [reply cmd/maya-apiserver/app/command/start.go]: doable.. its just another way.. followed from pool-mgmt code

File: cmd/maya-apiserver/cstor-operator/spc/runner.go (hunk 3 above)
Comment (vishnuitta) [reply cmd/maya-apiserver/cstor-operator/spc/runner.go]: Run is called from Cobra if I'm not wrong.. so, added here

File: cmd/maya-apiserver/cstor-operator/spc/runner.go [hunk 4]
```diff
@@ -40,13 +45,15 @@ func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
 	// Wait for the k8s caches to be synced before starting workers
 	glog.Info("Waiting for informer caches to sync")
 	if ok := cache.WaitForCacheSync(stopCh, c.spcSynced); !ok {
-		return fmt.Errorf("failed to wait for caches to sync")
+		err = fmt.Errorf("failed to wait for caches to sync")
```
Comment (prateekpandey14) [inline cmd/maya-apiserver/cstor-operator/spc/runner.go]: Why this change ?

File: cmd/maya-apiserver/cstor-operator/spc/runner.go [hunk 5]
```diff
@@ -40,13 +45,15 @@ func (c *Controller) Run(threadiness int, stopCh <-chan struct{}) error {
...
+		return err
 	}
 
 	glog.Info("Checking for preupgrade tasks")
-	err := c.performPreupgradeTasks()
+	err = c.performPreupgradeTasks()
 	if err != nil {
-		return fmt.Errorf("failure in preupgrade tasks: %v", err)
+		err = fmt.Errorf("failure in preupgrade tasks: %v", err)
```
Comment (prateekpandey14) [inline cmd/maya-apiserver/cstor-operator/spc/runner.go]: Why this change ?

File: cmd/maya-apiserver/app/command/start.go (hunk 1 above)
Comment (vishnuitta) [reply cmd/maya-apiserver/app/command/start.go]: there is errors package of maya which doesn't work for wrapping..

Comment (vishnuitta) [reply cmd/maya-apiserver/app/command/start.go]: so, using fmt..

Comment (vishnuitta) [reply cmd/maya-apiserver/app/command/start.go]: ok.. instead of wrap.. will use errors.Errorf itself

File: cmd/maya-apiserver/app/command/start.go (hunk 2 above)
Comment (mittachaitu) [reply cmd/maya-apiserver/app/command/start.go]: Yeah after looking into it I also resolved the comment
